File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    private static int answer = 0;
    
    private void dfs(int[] numbers, int target, int depth, int total) {
        if (depth == numbers.length) {
            if (total == target) {
                answer++;
            }
            return;
        }
        
        dfs(numbers, target, depth+1, total+numbers[depth]);
        dfs(numbers, target, depth+1, total-numbers[depth]);
    }
    
    public int solution(int[] numbers, int target) {        
        dfs(numbers, target, 0, 0);
        
        return answer;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    global answer
    answer = 0
    
    def dfs(idx, cnt, total):
        global answer 
        
        if cnt == len(numbers):
            if total == target:
                answer += 1
            return
        dfs(idx + 1, cnt + 1, total + numbers[idx])
        dfs(idx + 1, cnt + 1, total - numbers[idx])
    
    dfs(0, 0, 0)
    
    return answer